src/librustc/mir/interpret/value.rs

```diff
@@ -25,57 +26,130 @@ pub enum ConstValue<'tcx> {
-            ConstValue::Scalar(val) => Some(val),
+    pub fn try_as_by_ref(&self) -> Option<(AllocId, &'tcx Allocation, Size)> {
+        match self {
+            ConstValue::Unevaluated(..) => None,
+            ConstValue::ByRef(a, b, c) => Some((*a, *b, *c)),
         }
     }
 
     #[inline]
-    pub fn try_to_bits(&self, size: Size) -> Option<u128> {
-        self.try_to_scalar()?.to_bits(size).ok()
+    /// if this is ByRef, return the same thing but with the offset increased by `n`
+    pub fn try_offset(&self, n: Size) -> Option<Self> {
+        let (id, alloc, offset) = self.try_as_by_ref()?;
+        Some(ConstValue::ByRef(id, alloc, offset + n))
+    }
+
+    #[inline]
+    pub fn try_get_bytes(&self, hdl: impl HasDataLayout, n: Size, align: Align) -> Option<&[u8]> {
+        let (_, alloc, offset) = self.try_as_by_ref()?;
+        alloc.get_bytes(hdl, offset, n, align).ok()
+    }
+
+    #[inline]
+    pub fn try_to_bits(&self, hdl: impl HasDataLayout, layout: TyLayout<'tcx>) -> Option<u128> {
+        let bytes = self.try_get_bytes(hdl, layout.size, layout.align)?;
+        let endian = hdl.data_layout().endian;
+        super::read_target_uint(endian, &bytes).ok()
     }
 
     #[inline]
-    pub fn try_to_ptr(&self) -> Option<Pointer> {
-        self.try_to_scalar()?.to_ptr().ok()
+    pub fn try_to_usize(&self, hdl: impl HasDataLayout) -> Option<u128> {
+        let size = hdl.data_layout().pointer_size;
+        let align = hdl.data_layout().pointer_align;
+        let bytes = self.try_get_bytes(hdl, size, align)?;
+        let endian = hdl.data_layout().endian;
+        super::read_target_uint(endian, &bytes).ok()
+    }
+
+    #[inline]
+    pub fn try_to_ptr(
+        &self,
+        hdl: impl HasDataLayout,
+    ) -> Option<Pointer> {
+        let (_, alloc, offset) = self.try_as_by_ref()?;
+        alloc.read_scalar(hdl, offset).ok()?.to_ptr().ok()
+    }
+
+    /// e.g. for vtables, fat pointers or single pointers
+    #[inline]
+    pub fn new_pointer_list(
+        list: &[Scalar],
+        tcx: ty::TyCtxt<'_, '_, 'tcx>,
+    ) -> Self {
+        let ps = tcx.data_layout().pointer_size;
+        let mut alloc = Allocation::undef(
+            ps * list.len() as u64,
+            tcx.data_layout().pointer_align,
+        );
+        alloc.undef_mask.set_range_inbounds(Size::ZERO, ps * list.len() as u64, true);
+        for (i, s) in list.iter().enumerate() {
+            let (int, ptr) = match s {
+                Scalar::Bits { bits, size } => {
+                    assert!(*size as u64 == ps.bytes());
+                    (*bits as u64, None)
+                }
+                Scalar::Ptr(ptr) => (ptr.offset.bytes(), Some(ptr)),
+            };
+            let i = i * ps.bytes() as usize;
+            let j = i + ps.bytes() as usize;
+            super::write_target_uint(
+                tcx.data_layout().endian,
+                &mut alloc.bytes[i..j],
+                int.into(),
+            ).unwrap();
```

Comment:
Uh, so here we are duplicating `write_sclar`? Not good. You moved so many methods here, including `read_scalar`, why not also `write_scalar`?